fastfield_codecs/src/serialize.rs

```diff
@@ -21,107 +21,153 @@ use std::io;
+        } else {
+            VInt(0u64).serialize(writer)?;
+        }
+        self.codec_type.serialize(writer)?;
+        Ok(())
+    }
+
+    fn deserialize<R: io::Read>(reader: &mut R) -> io::Result<Self> {
+        let num_vals = VInt::deserialize(reader)?.0;
+        let min_value = VInt::deserialize(reader)?.0;
+        let amplitude = VInt::deserialize(reader)?.0;
+        let max_value = min_value + amplitude;
+        let gcd_u64 = VInt::deserialize(reader)?.0;
+        let codec_type = FastFieldCodecType::deserialize(reader)?;
+        Ok(Header {
+            num_vals,
+            min_value,
+            max_value,
+            gcd: NonZeroU64::new(gcd_u64),
+            codec_type,
+        })
     }
-    Some(GCDParams {
-        gcd,
-        min_value,
-        num_vals: column.num_vals(),
-    })
 }
 
 /// Returns correct the reader wrapped in the `DynamicFastFieldReader` enum for the data.
 pub fn open<T: MonotonicallyMappableToU64>(
     mut bytes: OwnedBytes,
 ) -> io::Result<Arc<dyn Column<T>>> {
-    let codec_type = FastFieldCodecType::deserialize(&mut bytes)?;
-    open_from_id(bytes, codec_type)
-}
-
-fn open_codec_from_bytes<C: FastFieldCodec, Item: MonotonicallyMappableToU64>(
-    bytes: OwnedBytes,
-) -> io::Result<Arc<dyn Column<Item>>> {
-    let reader = C::open_from_bytes(bytes)?;
-    Ok(Arc::new(monotonic_map_column(reader, Item::from_u64)))
-}
-
-pub fn open_gcd_from_bytes<WrappedCodec: FastFieldCodec>(
-    bytes: OwnedBytes,
-) -> io::Result<impl Column> {
-    let footer_offset = bytes.len() - 24;
-    let (body, mut footer) = bytes.split(footer_offset);
-    let gcd_params = GCDParams::deserialize(&mut footer)?;
-    let gcd_remap = move |val: u64| gcd_params.min_value + gcd_params.gcd * val;
-    let reader: WrappedCodec::Reader = WrappedCodec::open_from_bytes(body)?;
-    Ok(monotonic_map_column(reader, gcd_remap))
+    let header = Header::deserialize(&mut bytes)?;
+    match header.codec_type {
+        FastFieldCodecType::Bitpacked => open_specific_codec::<BitpackedCodec, _>(bytes, &header),
+        FastFieldCodecType::Linear => open_specific_codec::<LinearCodec, _>(bytes, &header),
+        FastFieldCodecType::BlockwiseLinear => {
+            open_specific_codec::<BlockwiseLinearCodec, _>(bytes, &header)
+        }
+    }
 }
 
-fn open_codec_with_gcd<C: FastFieldCodec, Item: MonotonicallyMappableToU64>(
+fn open_specific_codec<C: FastFieldCodec, Item: MonotonicallyMappableToU64>(
     bytes: OwnedBytes,
+    header: &Header,
 ) -> io::Result<Arc<dyn Column<Item>>> {
-    let reader = open_gcd_from_bytes::<C>(bytes)?;
-    Ok(Arc::new(monotonic_map_column(reader, Item::from_u64)))
+    let normalized_header = header.normalized();
+    let reader = C::open_from_bytes(bytes, normalized_header)?;
+    let min_value = header.min_value;
+    if let Some(gcd) = header.gcd {
+        let monotonic_mapping = move |val: u64| Item::from_u64(min_value + val * gcd.get());
```

Comment:
how the stats like `min_value` are handled is a little opaque now. We could use the common header for every fastfield as datasource, instead of wrapping and computation